Ipa interface overview (#1)
KevinUTAT/oneAPI-samples_wenkaixu_fork: DirectProgramming/C++SYCL_FPGA/Tutorials/Features/ip_authoring_interfaces_overview/csr-pipes/Readme.md

# CSR Data
This implementation uses a register-mapped invocation interface, and demonstrates how to use host pipes to return output into the CSR using the `protocol_avalon_mm` pipe protocol.

![](../assets/csr_out.svg)

## Invocation Interface
By default, an un-decorated oneAPI kernel will have all its control signals and arguments mapped into the IP component's control/status register (CSR).

## Data Interface - CSR Pipe

> In this design, input `InputPipeA` and `InputPipeB` are implemented as host pipes, as in [Streaming Data](host-pipes). However, the output `OutputPipeC` is implemented to write a single vector sum to the IP component's CSR.

Reviewer: ```suggestion
In this design, input `InputPipeA` and `InputPipeB` are implemented as host pipes, as in [Streaming Data](../host-pipes/). However, the output `OutputPipeC` is implemented to write a single vector sum to the IP component's CSR.
```

# CSR Data
This implementation uses a register-mapped invocation interface, and demonstrates how to use host pipes to return output into the CSR using the `protocol_avalon_mm` pipe protocol.

![](../assets/csr_out.svg)

## Invocation Interface
By default, an un-decorated oneAPI kernel will have all its control signals and arguments mapped into the IP component's control/status register (CSR).

## Data Interface - CSR Pipe

> In this design, input `InputPipeA` and `InputPipeB` are implemented as host pipes, as in [Streaming Data](../host-pipes/). However, the output `OutputPipeC` is implemented to write a single vector sum to the IP component's CSR.

Reviewer 2: ```suggestion
In this design, the inputs `InputPipeA` and `InputPipeB` are implemented as streaming interfaces using pipes, as in [Streaming Data](../host-pipes/). However, the output `OutputPipeC` is implemented to write a single vector sum to the IP component's CSR.
```

avoid saying 'host pipes'. I won't correct this anymore, but please ctrl + f for 'host pipes' in the rest of these readmes

> To configure a host pipe to map to the CSR, set its protocol as `AVALON_MM` or `AVALON_MM_USES_READY` (from `sycl::ext::intel::prototype::internal` namespace). See how to [specify pipe protocol](../host-pipes/Readme.md#Host-Pipe-Properties).

Reviewer 3: i think these protocol names and namespaces have changed. double check to make sure they're up to date.

> To configure a host pipe to map to the CSR, set its protocol as `protocol_name::avalon_mm` or `protocol_name::avalon_streaming_uses_ready` (from `sycl::ext::intel::experimental` namespace). See how to [specify pipe protocol](../host-pipes/Readme.md#Host-Pipe-Properties).

Reviewer 2: ```suggestion
To configure a host pipe to map to the CSR, set its protocol as `protocol_name::avalon_mm` or `protocol_name::avalon_streaming_uses_ready` (from `sycl::ext::intel::experimental` namespace). 
```

Alternatively, you can use the follwing property shorthands:
- `sycl::ext::intel::experimental::protocol_avalon_mm_uses_ready`
- `sycl::ext::intel::experimental::protocol_avalon_mm`

For more detail on how to use host pipes in general, refer to the [Streaming Data](../host-pipes/) part of this code sample.

## Build the Design

> This design supports four compilation options: Emulator, Simulator, Optimization Report, FPGA Hardware. For details on the different compilation options, see [fpga_compile](https://github.com/oneapi-src/oneAPI-samples/tree/master/DirectProgramming/C%2B%2BSYCL_FPGA/Tutorials/GettingStarted/fpga_compile) tutorial.

Reviewer 3: ```suggestion
This design supports four compilation options: Emulator, Simulator, Optimization Report, FPGA Hardware. For details on the different compilation options, see the [fpga_compile](https://github.com/oneapi-src/oneAPI-samples/tree/master/DirectProgramming/C%2B%2BSYCL_FPGA/Tutorials/GettingStarted/fpga_compile) tutorial.
```

>
> For more information on configuring environment variables, see [Use the setvars Script with Linux* or macOS*](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-linux-or-macos.html) or [Use the setvars Script with Windows*](https://www.intel.com/content/www/us/en/develop/documentation/oneapi-programming-guide/top/oneapi-development-environment-setup/use-the-setvars-script-with-windows.html).

### On Linux*

1. Change to the sample directory.
2. Configure the build system for the Agilex™ 7 device family, which is the default.

   ```
   mkdir build
   cd build
   cmake ..
   ```

   > **Note**: You can change the default target by using the command:
   >  ```
   >  cmake .. -DFPGA_DEVICE=<FPGA device family or FPGA part number>
   >  ``` 

3. Compile the design. (The provided targets match the recommended development flow.)

   1. Compile for emulation (fast compile time, targets emulated FPGA device).
      ```
      make fpga_emu
      ```

   2. Compile for simulation (fast compile time, targets simulator FPGA device):
      ```
      make fpga_sim
      ```

>    3. Generate HTML performance report.

Reviewer 2: @yuguen-intel why do we put 'simulation' before the report? simulation takes longer to compile than the report.

Reviewer 4: I don't have a preference

   4. Compile with Quartus place and route (To get accurate area estimate, longer compile time).
      ```
      nmake fpga
      ```

>**Note**: If you encounter any issues with long paths when compiling under Windows*, you may have to create your ‘build’ directory in a shorter path, for example `C:\samples\build`. You can then run cmake from that directory, and provide cmake with the full path to your sample directory.

## Run the Design

### On Linux

#### Run on FPGA Emulator

1. Run the sample on the FPGA emulator (the kernel executes on the CPU).
   ```
   ./vector_add.fpga_emu
   ```

#### Run on FPGA Simulator

1. Run the sample on the FPGA simulator.
   ```
   CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=1 ./vector_add.fpga_sim
   ```

### On Windows

#### Run on FPGA Emulator

1. Run the sample on the FPGA emulator (the kernel executes on the CPU).
   ```
   vector_add.fpga_emu.exe
   ```

#### Run on FPGA Simulator

1. Run the sample on the FPGA simulator.
   ```
   set CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=1
   vector_add.fpga_sim.exe
   set CL_CONTEXT_MPSIM_DEVICE_INTELFPGA=
   ```

## Example Output

```
add two vectors of size 256
PASSED
```

> You can found the address of the output CSR in `vector_add.report.prj\include\kernel_headers\SimpleVAddPipes_register_map.hpp`:

Reviewer 2: ```suggestion
You can find the address of the output CSR in `vector_add.report.prj\include\kernel_headers\SimpleVAddPipes_register_map.hpp`:
```
